### PR TITLE
Implement Pipelining

### PR DESCRIPTION
**Pull request by ihsandemir**
Implemented Pipelining the same as Java client implementation at https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/core/Pipelining.java.

### CODE REVIEW

**Comment by ihsandemir** [comment]
verify

**Comment by ihsandemir** [comment]
verify

**Comment by ihsandemir** [comment]
verify

File: hazelcast/include/hazelcast/client/Pipelining.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+             * Adds a future to this Pipelining or blocks until there is capacity to add the future to the Pipelining.
+             * <p>
+             * This call blocks until there is space in the Pipelining, but it doesn't mean that the invocation that
+             * returned the ICompletableFuture got blocked.
+             *
+             * @param future the future to add.
+             * @return the future added.
+             * @throws NullPointerException if future is null.
+             */
+            const boost::shared_ptr<ICompletableFuture<E> >
```
Comment (sancar) [inline hazelcast/include/hazelcast/client/Pipelining.h]: We should either return non-const shared ptr, or const shrd_ptr & . 
Returning const and copy does not cause any problem but it is confusing.

Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/Pipelining.h]: OK, changed to `const boost::shared_ptr<ICompletableFuture<E> > &`

File: hazelcast/include/hazelcast/client/Pipelining.h [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            class PipeliningExecutionCallback : public ExecutionCallback<E> {
+            public:
+                PipeliningExecutionCallback(const boost::shared_ptr<Pipelining> &pipelining) : pipelining(pipelining) {}
+
+                virtual void onResponse(const boost::shared_ptr<E> &response) {
+                    pipelining->up();
+                }
+
+                virtual void onFailure(const boost::shared_ptr<exception::IException> &e) {
+                    pipelining->down();
```
Comment (sancar) [inline hazelcast/include/hazelcast/client/Pipelining.h]: This should be `up`

Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/Pipelining.h]: good catch

**Comment by ihsandemir** [comment]
verify

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

**Pull request merged**